slatedb/src/manifest/store.rs

```diff
@@ -27,6 +27,324 @@ use std::sync::Arc;
+        &self.record
+    }
+
+    fn next_id(&self) -> u64 {
+        self.id + 1
+    }
+
+    pub(crate) async fn refresh(&mut self) -> Result<&T, SlateDBError> {
+        let Some((id, new_val)) = self.store.try_read_latest().await? else {
+            return Err(InvalidDBState);
+        };
+        self.id = id;
+        self.record = new_val;
+        Ok(&self.record)
+    }
+
+    pub(crate) async fn update_dirty(&mut self, dirty: DirtyRecord<T>) -> Result<(), SlateDBError> {
+        if dirty.id != self.id {
+            return Err(ManifestVersionExists);
+        }
+        let next = self.next_id();
+        self.store.write(next, &dirty.value).await?;
+        self.id = next;
+        self.record = dirty.value;
+        Ok(())
+    }
+
+    pub(crate) async fn maybe_apply_update<F>(&mut self, mutator: F) -> Result<(), SlateDBError>
+    where
+        F: Fn(&StoredRecord<T>) -> Result<Option<DirtyRecord<T>>, SlateDBError>,
+    {
+        loop {
+            let Some(dirty) = mutator(self)? else {
+                return Ok(());
+            };
+            match self.update_dirty(dirty).await {
+                Err(SlateDBError::ManifestVersionExists) => {
+                    self.refresh().await?;
+                    continue;
+                }
+                Err(e) => return Err(e),
+                Ok(()) => return Ok(()),
+            }
+        }
+    }
+}
+
+// Local mutable view of a versioned record
+#[derive(Clone, Debug)]
+pub(crate) struct DirtyRecord<T> {
+    id: u64,
+    value: T,
+}
+
+impl<T> DirtyRecord<T> {
+    #[allow(dead_code)]
+    fn id(&self) -> u64 {
+        self.id
+    }
+    #[allow(dead_code)]
+    fn into_value(self) -> T {
+        self.value
+    }
+}
+
+// Generic fenceable wrapper using epoch getters/setters
+pub(crate) struct FenceableRecord<T: Clone> {
+    stored: StoredRecord<T>,
+    local_epoch: u64,
+    get_epoch: fn(&T) -> u64,
+    #[allow(dead_code)]
+    set_epoch: fn(&mut T, u64),
+}
+
+impl<T: Clone + Send + Sync> FenceableRecord<T> {
+    pub(crate) async fn init(
+        mut stored: StoredRecord<T>,
+        record_update_timeout: Duration,
+        system_clock: Arc<dyn SystemClock>,
+        get_epoch: fn(&T) -> u64,
+        set_epoch: fn(&mut T, u64),
+    ) -> Result<Self, SlateDBError> {
+        utils::timeout(
+            system_clock,
+            record_update_timeout,
+            "record update",
+            async {
+                loop {
+                    let local_epoch = get_epoch(stored.record()) + 1;
+                    let mut new_val = stored.record().clone();
+                    set_epoch(&mut new_val, local_epoch);
+                    let dirty = DirtyRecord {
+                        id: stored.id(),
+                        value: new_val,
+                    };
+                    match stored.update_dirty(dirty).await {
+                        Err(ManifestVersionExists) => {
+                            stored.refresh().await?;
```

Comment:
We had this comment:

```
// The manifest may have been updated by a reader, or
// we may have gotten this error after successfully updating
// if we failed to get the response. Either way, refresh
// the manifest and try the bump again.
```

Can you keep it? (but replace "manifest" with "file" etc)